src/node_jsvmapi.cc

```diff
@@ -465,22 +465,58 @@ namespace v8impl {
 
 }  // end of namespace v8impl
 
-// We define a macro here because, with full error handling, we may want to
-// perform additional actions at the top of our public APIs, such as bail,
-// returning an error code at the very top of each public API in the case that
-// there's already an exception pending in v8impl::TryCatch::lastException
+   // We define a macro here because, with full error handling, we may want to
+   // perform additional actions at the top of our public APIs, such as bail,
+   // returning an error code at the very top of each public API in the case that
+   // there's already an exception pending in v8impl::TryCatch::lastException
 #define NAPI_PREAMBLE(e) \
   v8impl::TryCatch tryCatch(v8impl::V8IsolateFromJsEnv((e)))
 
-napi_env napi_get_current_env() {
-  return v8impl::JsEnvFromV8Isolate(v8::Isolate::GetCurrent());
+   // Static last error returned from napi_get_last_error_info
+napi_extended_error_info static_last_error;
+
+// Warning: Keep in-sync with napi_errorvalue enum
+const char* error_messages[] = {
+  NULL,
+  "Invalid pointer passed as argument",
+  "An object was expected",
+  "A string was expected",
+  "Unknown failure"
+};
+
+const napi_extended_error_info* napi_get_last_error_info() {
+  // Wait until someone requests the last error information to fetch the error message string
+  static_last_error.error_message = error_messages[static_last_error.error_code];
+
+  return &static_last_error;
 }
 
-napi_value napi_create_function(
+napi_errorvalue napi_set_last_error_code(napi_errorvalue error_code) {
+  static_last_error.error_code = error_code;
+
+  return error_code;
+}
+
+napi_errorvalue napi_get_current_env(napi_env* e) {
+  if (!e) {
+    return napi_set_last_error_code(napi_invalid_arg);
+  }
+
+  *e = v8impl::JsEnvFromV8Isolate(v8::Isolate::GetCurrent());
+  return napi_ok;
+}
+
+napi_errorvalue napi_create_function(
     napi_env e,
     napi_callback cb,
-    void* data) {
+    void* data,
+    napi_value* result) {
   NAPI_PREAMBLE(e);
+
+  if (!result) {
+    return napi_set_last_error_code(napi_invalid_arg);
+  }
```

Comment:
Should there be a macro for this check, since it's repeated for nearly every API?